core/templates/services/playthrough.service.ts

```diff
@@ -1,334 +1,359 @@
+    if (this.currStateName() === destStateName) {
+      return;
+    }
+    if (this.pathOfVisitedStates.includes(destStateName)) {
+      const cycleOfVisitedStates = (
+        this.makeCycle(this.pathOfVisitedStates.indexOf(destStateName)));
+      if (angular.equals(this.cycleOfVisitedStates, cycleOfVisitedStates)) {
+        this.numLoops += 1;
+      } else {
+        this.cycleOfVisitedStates = cycleOfVisitedStates;
+        this.numLoops = 1;
+      }
+       this.pathOfVisitedStates.length = 0;
+    }
+    this.pathOfVisitedStates.push(destStateName);
+  }
+
+  generateIssueCustomizationArgs(): CyclicStateTransitionsCustomizationArgs {
+    return {
+      state_names: {value: this.cycleOfVisitedStates}
+    };
+  }
+
+  private makeCycle(collisionIndex: number): string[] {
+    const collision = this.pathOfVisitedStates[collisionIndex];
+    const cycleWithNoCollision =
+    this.pathOfVisitedStates.slice(collisionIndex);
+    return [...cycleWithNoCollision, collision];
+  }
+
+  private currStateName(): string {
+    return this.pathOfVisitedStates[this.pathOfVisitedStates.length - 1];
+  }
+}
+
+class EarlyQuitTracker {
+  private stateName: string = ' ';
+  private expDurationInSecs: number | null = null;
+
+  foundAnIssue(): boolean {
+    return (
+      this.expDurationInSecs !== null &&
+      this.expDurationInSecs < ServicesConstants.EARLY_QUIT_THRESHOLD_IN_SECS);
+  }
+
+  recordExplorationQuit(
+      stateName: string,
+      expDurationInSecs: number | null): void {
+    if (this.stateName !== null && this.expDurationInSecs !== null) {
+      this.stateName = stateName;
+      this.expDurationInSecs = expDurationInSecs;
+    }
+  }
+
+
+  generateIssueCustomizationArgs(): EarlyQuitCustomizationArgs {
+    if (this.expDurationInSecs !== null) {
+      return {
+        state_name!: {value: this.stateName},
+        time_spent_in_exp_in_msecs!: {value: this.expDurationInSecs * 1000},
+      };
+    } else {
+      throw console.error('Value should not be null !');
+    }
+  }
+}
+
+class MultipleIncorrectAnswersTracker {
+  private currStateName: string;
+  private numTries: number;
+
+  constructor(initStateName: string) {
+    this.currStateName = initStateName;
+    this.numTries = 0;
+  }
+
+  foundAnIssue(): boolean {
+    return this.numTries >= ServicesConstants.NUM_INCORRECT_ANSWERS_THRESHOLD;
+  }
+
+  recordStateTransition(destStateName: string): void {
+    if (this.currStateName === destStateName) {
+      this.numTries += 1;
+    } else {
+      this.currStateName = destStateName;
+      this.numTries = 0;
+    }
+  }
+
+  generateIssueCustomizationArgs(
+  ): MultipleIncorrectSubmissionsCustomizationArgs {
+    return {
+      state_name: {value: this.currStateName},
+      num_times_answered_incorrectly: {value: this.numTries},
+    };
+  }
+}
+
+ @Injectable({
+   providedIn: 'root'
+ })
+export class PlaythroughService {
+  private explorationId: string = ' ';

```

Comment:
Why is this set to empty string?

Reply:
Since it has to be intialised with something and we cant assign it null explicitely as it gives an strict type error.

Reply:
ditto as above